packages/xsnap/src/avaAssertXS.js

```diff
@@ -0,0 +1,351 @@
+
+  if (!theHarness) {
+    theHarness = it;
+  }
+  return it;
+}
+
+/**
+ * @param {*} exc
+ * @param {Expectation} expectation
+ *
+ * @typedef {{ instanceOf: Function } | { message: string | RegExp }=} Expectation
+ */
+function checkExpectation(exc, expectation) {
+  if (!expectation) return true;
+  if ('instanceOf' in expectation) return exc instanceof expectation.instanceOf;
+  if ('message' in expectation) {
+    const { message } = expectation;
+    return typeof message === 'string'
+      ? exc.message === message
+      : exc.message.match(message);
+  }
+  throw Error(`not implemented: ${JSON.stringify(expectation)}`);
+}
+
+/**
+ * Emulate ava assertion API
+ *
+ * ref https://github.com/avajs/ava/blob/main/docs/03-assertions.md
+ *
+ * @param {Harness} htest
+ * @param {TapFormat} out
+ *
+ * @typedef {ReturnType<typeof makeTester>} Tester
+ */
+function makeTester(htest, out) {
+  /** @type {number?} */
+  let pending;
+
+  /** @type {(r: boolean, info?: string) => void} */
+  function assert(result, info) {
+    if (typeof pending === 'number') {
+      pending -= 1;
+    }
+    const testNum = htest.finish(result);
+    if (result) {
+      out.ok(testNum, info);
+    } else {
+      out.notOk(testNum, info);
+    }
+  }
+
+  function truthy(/** @type {unknown} */ value, msg = 'should be truthy') {
+    assert(!!value, msg);
+  }
+
+  /** @type {(a: unknown, e: unknown) => void } */
+  function deepEqTest(actual, expected) {
+    try {
+      assert(deepEqual(actual, expected), 'should be deep equal');
+    } catch (reason) {
+      const summary = JSON.stringify({ actual, expected });
+      assert(false, `should be deep equal: ${summary} : ${reason.message}`);
+    }
+  }
+
+  const t = freeze({
+    plan(/** @type {number} */ count) {
+      pending = count;
+    },
+    get pending() {
+      return pending;
+    },
+    get context() {
+      return htest.context;
+    },
+    pass(/** @type {string} */ message) {
+      assert(true, message);
+    },
+    fail(/** @type {string} */ message) {
+      assert(false, message);
+    },
+    assert,
+    truthy,
+    falsy(/** @type {unknown} */ value, message = 'should be falsy') {
+      assert(!value, message);
+    },
+    true(/** @type {unknown} */ value, message = 'should be true') {
+      assert(value === true, message);
+    },
+    false(/** @type {unknown} */ value, message = 'should be false') {
+      assert(value === false, message);
+    },
+    /** @type {(a: unknown, b: unknown, m?: string) => void} */
+    is(a, b, message = 'should be identical') {
+      assert(Object.is(a, b), message);
+    },
+    /** @type {(a: unknown, b: unknown, m?: string) => void} */
+    not(a, b, message = 'should not be identical') {
+      assert(!Object.is(a, b), message);
+    },
+    deepEqual: deepEqTest,
+    /** @type {(a: unknown, b: unknown, m?: string) => void} */
+    notDeepEqual(a, b, message = 'should not be deep equal') {
+      assert(!deepEqual(a, b), message);
```

Comment:
Aha! Because `deepEqual` errors on some input, this will still throw on that input. IIUC `t.notDeepEqual([], [8])` will throw because the lengths are different. Yes? Is that what you want? What does ava do? What I'd expect that test to do is succeed.

Reply:
> Is that what you want?

no. good catch.